### PR TITLE
Create pascals-triangle.java

### PR DESCRIPTION
**Pull request by soyeon51696**

### CODE REVIEW

File: study1/week3/pascals-triangle.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class Solution {
+    public List<List<Integer>> generate(int numRows) {
+        //When numRows is n, time complexity is O(n^n)
+        //Because, the first for loop -> visits n times and the inner for-loop -> visits up to n times 
+        //(because j repeats at most i times, if i is n then j also n)
+        
+        //When numRows is n, space complexity is O(n)
+        //Because, n lists are stored in result list, so the length of the result list is n.
+        //And then in also inner for-loop n lists are stored in semiResult list.
+        //So O(n + n) = O(2n) = O(n)
```
Comment (h-joo) [inline study1/week3/pascals-triangle.java]: How many elements will variable `result` store?

File: study1/week3/pascals-triangle.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                    semiResult.add(1);
+                }
+                else if (j == i) {
+                    semiResult.add(1);
+                }
+                else {
+                    semiResult.add(result.get(i - 1).get(j - 1) + result.get(i - 1).get(j));
+                }
+            }
+            result.add(semiResult);
```
Comment (h-joo) [inline study1/week3/pascals-triangle.java:30]: Optional: You can also move this to line 18 as the semiResult keeps a reference to the list and it doesn't matter adding it before or after the loop on line 19.

File: study1/week3/pascals-triangle.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        //And then in also inner for-loop n lists are stored in semiResult list.
+        //So O(n + n) = O(2n) = O(n)
+      
+        List<List<Integer>> result = new ArrayList<>();
+
+        for (int i = 0; i < numRows; i++) { //until numRow
+            List<Integer> semiResult = new ArrayList<>();
+
+            for (int j = 0; j <= i; j++) { //j <= i -> 1st -> 1 element, 2nd -> 2 elements, 3rd -> 3 elemets, ...
+                if (j == 0) {
```
Comment (h-joo) [inline study1/week3/pascals-triangle.java]: `j == 0` is for the first element and `j == i` is for the last element, so you can move these outside of the loop and reduce the number of comparisons. so doing the following would simplify and reduce comparisons: 
```java
semiResult.add(1);
for (int j = 1; j < i; j++) {
   semiResult.add(result.get(i - 1).get(j - 1) + result.get(i - 1).get(j));
}
semiResult.add(1);
```

**Review by h-joo — APPROVED** [review]

**Comment by h-joo** [comment]
Please convert this to a non-draft PR and merge :)

File: study1/week3/pascals-triangle.java (hunk 1 above)
Comment (soyeon51696) [reply study1/week3/pascals-triangle.java]: numRows lists?.. and numRow lists store also n lists
so is space complexity O(n^2)?

**Review by h-joo — APPROVED** [review]

**Pull request merged**